extension/src/time_weighted_average.rs

```diff
@@ -0,0 +1,422 @@
+                (Some(state), None) => Some(state),
+                (Some(mut state), Some(next)) =>  {
+                    let next = TimeWeightTransState{combine_buffer:vec![next.to_TimeWeightSummary()], buffer: vec![], method: Some(*next.method)};
+                    state.combine_push(&next); 
+                    Some(state.into())
+                },
+            }
+        })
+    }
+}
+
+
+#[pg_extern(schema = "timescale_analytics_experimental")]
+pub fn time_weight_combine(
+    state1: Option<Internal<TimeWeightTransState>>,
+    state2: Option<Internal<TimeWeightTransState>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+)  -> Option<Internal<TimeWeightTransState>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state1, state2) {
+                (None, None) => None,
+                (None, Some(state2)) => {let mut s = state2.clone(); s.sort_and_calc(); Some(s.into())},
+                (Some(state1), None) => {let mut s = state1.clone(); s.sort_and_calc(); Some(s.into())}, //should I make these return themselves?
+                (Some(state1), Some(state2)) => {
+                    let mut s1 = state1.clone(); // is there a way to avoid if it doesn't need it?
+                    s1.sort_and_calc(); 
+                    let mut s2 = state2.clone();
+                    s2.sort_and_calc();
+                    s2.combine_push(&s1);
+                    Some(s2.into())
+                }
+            }
+        })
+    }
+}
+
+#[pg_extern(schema = "timescale_analytics_experimental")]
+fn time_weight_final(
+    state: Option<Internal<TimeWeightTransState>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<time_weight_summary<'static>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let mut state = match state {
+                None => return None,
+                Some(state) => state.clone(),
+            };
+            state.sort_and_calc();
+            state.combine_sort_and_calc();
+            debug_assert!(state.combine_buffer.len() <= 1);
+            match state.combine_buffer.pop() {
+                None => None,
+                Some(st) => Some(
+                    flatten!(
+                        time_weight_summary {
+                            method: &st.method,
+                            first: &st.first,
+                            last: &st.last,
+                            w_sum: &st.w_sum,
+                    }
+                ).into())
+            }
+        })
+    }
+}
+
+extension_sql!(r#"
+
+
+CREATE AGGREGATE time_weight(method text, ts timestamptz, value DOUBLE PRECISION)
+(
+    sfunc = time_weight_trans,
+    stype = internal,
+    finalfunc = time_weight_final,
+    combinefunc = time_weight_combine,
+    serialfunc = time_weight_trans_serialize,
+    deserialfunc = time_weight_trans_deserialize,
+    parallel = restricted
+);
+
+CREATE AGGREGATE time_weight(tws time_weight_summary)
```

Comment:
I'm not sure I understand how this will be used.  Is this to handle aggregation of the already aggregated data?

Reply:
Yes. I will leave some comments about this, but, if we don't already have this for tdigest and the other things, we shoudl absolutely add, it's a huge improvement in the UI for re-aggregation from Continuous aggregates cause I don't think the Timescale side solution for that is coming for a little while and having that is a huge improvement.